include/deal.II/integrators/advection.h

```diff
@@ -257,6 +349,139 @@ namespace LocalIntegrators
         }
     }
 
+
+
+    /**
+     * Scalar case: Residual for upwind flux at the boundary for weak advection operator. This is the
+     * value of the trial function at the outflow boundary and zero else:
+     * @f[
+     * a_{ij} = \int_{\partial\Omega}
+     * [\mathbf w\cdot\mathbf n]_-
+     * (u_i-g) v_j \, ds
+     * @f]
+     *
+     * Here, <i>u</i> is the finite element function whose values are given in the argument <tt>input</tt>.
+     * <i>g</i> is the inhomogenous boundary value in the argument <tt>data</tt>.
+     *
+     * The <tt>velocity</tt> is provided as a VectorSlice, having <tt>dim</tt>
+     * vectors, one for each velocity component. Each of the vectors must
+     * either have only a single entry, if the advection velocity is
+     * constant, or have an entry for each quadrature point.
+     *
+     * The finite element can have several components, in which case each
+     * component is advected by the same velocity.
+     */
+    template <int dim>
+    inline void
+    upwind_value_residual(
+      Vector<double> &result,
+      const FEValuesBase<dim> &fe,
+      const std::vector<double> &input,
+      const std::vector<double> &data,
+      const VectorSlice<const std::vector<std::vector<double> > > &velocity,
+      double factor = 1.)
+    {
+      const unsigned int n_dofs = fe.dofs_per_cell;
+
+      AssertDimension(input.size(), fe.n_quadrature_points);
+      AssertDimension(data.size(), fe.n_quadrature_points);
+
+      AssertDimension(velocity.size(), dim);
+      const unsigned int v_increment = (velocity[0].size() == 1) ? 0 : 1;
+      if (v_increment == 1)
+        {
+          AssertVectorVectorDimension(velocity, dim, fe.n_quadrature_points);
+        }
+
+
+      for (unsigned k=0; k<fe.n_quadrature_points; ++k)
+        {
+          const double dx = factor * fe.JxW(k);
+
+          double nv = 0.;
+          for (unsigned int d=0; d<dim; ++d)
+            nv += fe.normal_vector(k)[d] * velocity[d][k * v_increment];
+
+          if (nv > 0)
+            {
+              for (unsigned i=0; i<n_dofs; ++i)
+                {
+                  const double v= fe.shape_value(i,k);
+                  const double u= input[k];
+                  const double g= data[k];
+                  result(i) += dx * nv *(u-g)*v;
+                }
+            }
+        }
+    }
+
+
+
+    /**
+     * Vector-valued case: Residual for upwind flux at the boundary for weak advection operator. This is the
+     * value of the trial function at the outflow boundary and zero else:
+     * @f[
+     * a_{ij} = \int_{\partial\Omega}
+     * [\mathbf w\cdot\mathbf n]_-
+     * (u_i-g) v_j \, ds
+     * @f]
```

Comment:
The only thing that remains unclear to me is whether this should actually be the outflow part of the boundary. If I understand @guidokanschat right, this should have been the inflow part (where n * velocity < 0), and this is where I would intuitively set boundary conditions in an advection problem with upwind flux. (The same applies to the scalar case above). I might be wrong, so please check this, then we can merge the pull request.